fix for #834 adding logic to make porch API calls atomic

mansoor17syed opened the pull request.

liamfallon commented on pkg/engine/engine.go:
```diff
@@ -151,21 +151,44 @@ func (cad *cadEngine) CreatePackageRevision(ctx context.Context, repositoryObj *
+	// Setup rollback function in case of errors
+	rollback := func() {
+		// Convert draft to PackageRevision for deletion
+		if pkgRev, err := repo.ClosePackageRevisionDraft(ctx, draft, 0); err == nil {
```
We probably should amend the interface to allow deletion of drafts rather than having to close a draft so we can delete it, but it's much more work to do that so i think this is fine for now.

Would you be able to create a new issue to update the repository..PacakgeRevisionDraft interface to add a delete() function to the interface?

mansoor17syed replied:
So the proposed change seems fine but we should  create a new issue to track adding a delete() function to the repository.PackageRevisionDraft interface? This would be a cleaner solution in the future. ?

liamfallon replied:
Yes. That would be my suggestion.

liamfallon commented on pkg/engine/engine.go:
```diff
@@ -151,21 +151,44 @@ func (cad *cadEngine) CreatePackageRevision(ctx context.Context, repositoryObj *
+	// Close the draft
+	repoPkgRev, err := repo.ClosePackageRevisionDraft(ctx, draft, 0)
+	if err != nil {
+		rollback()
```
If this call to ClosePacakgeRevisionDraft errors out, won't the rollback() function always error out too because it calls ClosePackageRevisionDraft again?

liamfallon replied:
I don't think you can do anything because if the closePackageRevisionDraft fails it will probably fail again in the rollback() function. Its probably best to just return the error generated by the call at line 185

liamfallon commented:
We should add some unit tests and an e2e test to cover the case where a draft creation fails and the "rollback()" function is called.

mansoor17syed commented:
> We should add some unit tests and an e2e test to cover the case where a draft creation fails and the "rollback()" function is called.

What specific test cases should we add if you have any in mind
Should we focus on unit tests first or e2e tests?

I will work on it based on your feedback

liamfallon commented:
> > We should add some unit tests and an e2e test to cover the case where a draft creation fails and the "rollback()" function is called.
> 
> What specific test cases should we add if you have any in mind Should we focus on unit tests first or e2e tests?
> 
> I will work on it based on your feedback

I think it would be best to do the unit tests first and do the e2e tests later.

On the e2e tests, you could probably add a test using the TestDeleteDraft() test as a template:
https://github.com/nephio-project/porch/blob/7dc71bb135ee11d49afffbeada26029175b4c3c9/test/e2e/e2e_test.go#L1137

mansoor17syed commented:
Hi @liamfallon 

I've addressed all the review comments and added the unit tests as well as attempted to include some end-to-end validations. I have created the issue also as suggested [https://github.com/nephio-project/nephio/issues/909](url) However, the tests are passing fine in my local environment but seem to be failing in the current CI run.

Could you please suggest any pointers or things I should check? It would be really helpful!


![image](https://github.com/user-attachments/assets/69f1bc29-34e3-4719-9b43-9b26914bd78e)


Thanks a lot for your time and support!

mansoor17syed commented:
Hi @liamfallon ,

Any feedback for me here ?
Thanks

efiacor commented:
> Hi @liamfallon ,
> 
> Any feedback for me here ? Thanks

Looks like we may have to regenerate the mockery stuff.

panic: interface conversion: *engine.mockRepository is not repository.Repository: missing method Close [recovered]
	panic: interface conversion: *engine.mockRepository is not repository.Repository: missing method Close
	
https://github.com/nephio-project/porch/blob/main/default-mockery.mk#L30

liamfallon commented:
> > Hi @liamfallon ,
> > Any feedback for me here ? Thanks
> 
> Looks like we may have to regenerate the mockery stuff.
> 
> panic: interface conversion: *engine.mockRepository is not repository.Repository: missing method Close [recovered] panic: interface conversion: *engine.mockRepository is not repository.Repository: missing method Close
> 
> https://github.com/nephio-project/porch/blob/main/default-mockery.mk#L30

Yes, there-s a makefile target for that
```
make generate-mocks
```

liamfallon commented:
> Hi @liamfallon ,
> 
> Any feedback for me here ? Thanks

Hi @mansoor17syed ,

Apologies, I should have got back to you on this before now.

It looks good. If you can generate the mocks and check them in and get the tests working we should be good to merge this.

mansoor17syed commented:
Hi @liamfallon  and @efiacor 

Thank you for the suggestions. I applied the changes, and the output below appears to pass afterward as well. Am I missing something?


![image](https://github.com/user-attachments/assets/236c9433-c141-4d66-9194-1d398733ff25)

liamfallon commented:
Did you check in the code generated for the mocks? That could be why the tests are failing on the build.

liamfallon commented:
Hi @mansoor17syed, when I pull your PR and try to run it locally, one of the setup scripts is breaking. I don't think its anything to do with your PR. I'll talk to the team in the morning and try to get to the bottom if it.

```
% scripts/setup-dev-env.sh 

** Install kind cluster: porch-test

Cluster already exists.
Set kubectl context to "kind-porch-test"

** Install MetalLB

namespace/metallb-system unchanged
customresourcedefinition.apiextensions.k8s.io/bfdprofiles.metallb.io unchanged
customresourcedefinition.apiextensions.k8s.io/bgpadvertisements.metallb.io unchanged
customresourcedefinition.apiextensions.k8s.io/bgppeers.metallb.io configured
customresourcedefinition.apiextensions.k8s.io/communities.metallb.io unchanged
customresourcedefinition.apiextensions.k8s.io/ipaddresspools.metallb.io unchanged
customresourcedefinition.apiextensions.k8s.io/l2advertisements.metallb.io unchanged
customresourcedefinition.apiextensions.k8s.io/servicel2statuses.metallb.io unchanged
serviceaccount/controller unchanged
serviceaccount/speaker unchanged
role.rbac.authorization.k8s.io/controller unchanged
role.rbac.authorization.k8s.io/pod-lister unchanged
clusterrole.rbac.authorization.k8s.io/metallb-system:controller unchanged
clusterrole.rbac.authorization.k8s.io/metallb-system:speaker unchanged
rolebinding.rbac.authorization.k8s.io/controller unchanged
rolebinding.rbac.authorization.k8s.io/pod-lister unchanged
clusterrolebinding.rbac.authorization.k8s.io/metallb-system:controller unchanged
clusterrolebinding.rbac.authorization.k8s.io/metallb-system:speaker unchanged
configmap/metallb-excludel2 unchanged
secret/metallb-webhook-cert unchanged
service/metallb-webhook-service unchanged
deployment.apps/controller unchanged
daemonset.apps/speaker unchanged
validatingwebhookconfiguration.admissionregistration.k8s.io/metallb-webhook-configuration configured
Waiting for controller to become ready...
deployment.apps/controller condition met
ipaddresspool.metallb.io/example unchanged
l2advertisement.metallb.io/empty unchanged

** Install Gitea and init test repos

Error: package must have an upstream reference 
```

liamfallon commented:
In the meantime, if you could rebase your PR and push the rebased version, that would help a lot.

mansoor17syed commented:
Hey @liamfallon ,
Thanks for your feedback and support. I also did a rebase along with the some  additional changes for tests. After testing and validating, everything is now passing. Could you please help me review it?

liamfallon commented:
/approve

liamfallon commented:
Thanks @mansoor17syed for all your efforts. Let's get one more /approve on it and we can merge it.

efiacor commented:
/approve
